TUN: Handling for TEB PI payload on TUN Intf

Remove the ethernet header from the skb when a
TEB payload is received on a TUN type device indicated
by the PI information in the buffer.

Signed-off-by: Mandeep Rohilla <mrohilla@vyatta.att-mail.com>
---
 drivers/net/tun.c |    9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1321,8 +1321,13 @@ static ssize_t tun_get_user(struct tun_s
 			}
 		}
 
-		skb_reset_mac_header(skb);
-		skb->protocol = pi.proto;
+		if (pi.proto == htons(ETH_P_TEB)) {
+			skb->protocol = eth_type_trans(skb, tun->dev);
+		} else {
+			skb_reset_mac_header(skb);
+			skb->protocol = pi.proto;
+		}
+
 		skb->dev = tun->dev;
 		break;
 	case IFF_TAP:
